source/loader/layers/sanitizer/msan/msan_interceptor.cpp

```diff
@@ -0,0 +1,493 @@
+    }
+
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t MsanInterceptor::preLaunchKernel(ur_kernel_handle_t Kernel,
+                                             ur_queue_handle_t Queue,
+                                             USMLaunchInfo &LaunchInfo) {
+    auto Context = GetContext(Queue);
+    auto Device = GetDevice(Queue);
+    auto ContextInfo = getContextInfo(Context);
+    auto DeviceInfo = getDeviceInfo(Device);
+
+    ManagedQueue InternalQueue(Context, Device);
+    if (!InternalQueue) {
+        getContext()->logger.error("Failed to create internal queue");
+        return UR_RESULT_ERROR_INVALID_QUEUE;
+    }
+
+    UR_CALL(prepareLaunch(DeviceInfo, InternalQueue, Kernel, LaunchInfo));
+
+    UR_CALL(updateShadowMemory(ContextInfo, DeviceInfo, InternalQueue));
+
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t MsanInterceptor::postLaunchKernel(ur_kernel_handle_t Kernel,
+                                              ur_queue_handle_t Queue,
+                                              USMLaunchInfo &LaunchInfo) {
+    // FIXME: We must use block operation here, until we support urEventSetCallback
+    auto Result = getContext()->urDdiTable.Queue.pfnFinish(Queue);
+
+    if (Result == UR_RESULT_SUCCESS) {
+        const auto &Report = LaunchInfo.Data->Report;
+
+        if (!Report.Flag) {
+            return Result;
+        }
+
+        ReportUsesUninitializedValue(LaunchInfo.Data->Report, Kernel);
+
+        exitWithErrors();
+    }
+
+    return Result;
+}
+
+ur_result_t
+MsanInterceptor::enqueueAllocInfo(std::shared_ptr<DeviceInfo> &DeviceInfo,
+                                  ur_queue_handle_t Queue,
+                                  std::shared_ptr<MsanAllocInfo> &AI) {
+    return DeviceInfo->Shadow->EnqueuePoisonShadow(Queue, AI->AllocBegin,
+                                                   AI->AllocSize, 0xff);
+}
+
+ur_result_t
+MsanInterceptor::updateShadowMemory(std::shared_ptr<ContextInfo> &ContextInfo,
+                                    std::shared_ptr<DeviceInfo> &DeviceInfo,
+                                    ur_queue_handle_t Queue) {
+    auto &AllocInfos = ContextInfo->AllocInfosMap[DeviceInfo->Handle];
+    std::scoped_lock<ur_shared_mutex> Guard(AllocInfos.Mutex);
+
+    for (auto &AI : AllocInfos.List) {
+        UR_CALL(enqueueAllocInfo(DeviceInfo, Queue, AI));
+    }
+    AllocInfos.List.clear();
+
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t MsanInterceptor::registerProgram(ur_program_handle_t Program) {
+    ur_result_t Result = UR_RESULT_SUCCESS;
+
+    getContext()->logger.info("registerSpirKernels");
+    Result = registerSpirKernels(Program);
+    if (Result != UR_RESULT_SUCCESS) {
+        return Result;
+    }
+
+    return Result;
+}
+
+ur_result_t MsanInterceptor::unregisterProgram(ur_program_handle_t) {
+    return UR_RESULT_SUCCESS;
+}
+
+ur_result_t MsanInterceptor::registerSpirKernels(ur_program_handle_t Program) {
```

Comment:
This, and a few other functions in this file, are nearly straight copy-paste from asan interceptor.

I know there's time pressure now, but please create a high-priority ticket to refactor this post merge. Maybe there should be a common base class for sanitizers that contains the shared code?

Reply:
Sure, we noticed this problem as well, after intense period of feature developing, we will try to merge common codes to make it more easily maintain.